### PR TITLE
TINY-10636: Update Stylelint and remove Gulp

### PR DESCRIPTION
**Pull request by MitchC1999**
Related Ticket: TINY-10636, TINY-10116

Description of Changes:
* Update Stylelint and fixed related issues
* Replaced gulp with grunt in Oxide

Pre-checks:
* [x] Changelog entry added
* [x] <s>Tests have been added (if applicable)</s>
* [x] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [x] Milestone set
* [x] Docs ticket created (if applicable)

GitHub issues (if applicable):

### CODE REVIEW

File: modules/oxide/tasks/less-task.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+const less = require('less');
+const fs = require('node:fs');
+const path = require('path');
+const LessPluginPresetEnv = require('less-plugin-preset-env');
+
+const presetEnv = new LessPluginPresetEnv({
+  browsers: ['last 2 Safari versions', 'iOS 14.0', 'last 2 Chrome versions', 'Firefox ESR'],
```
Comment (TheSpyder) [inline modules/oxide/tasks/less-task.js:7]:
```suggestion
  browsers: ['last 2 Safari versions', 'last 2 iOS major versions', 'last 2 Chrome versions', 'Firefox ESR'],
```
I was wondering if this was possible, looks like it is:
https://github.com/browserslist/browserslist#full-list

Unfortunately it doesn't seem to be capable of "chrome extended stable", although that changes frequently enough that we should be ok.

File: package.json [hunk 2]
```diff
@@ -38,53 +38,60 @@
...
-    "less-plugin-autoprefix": "^2.0.0",
+    "less": "^4.2.0",
+    "less-plugin-preset-env": "^1.2.6",
     "load-grunt-tasks": "^5.1.0",
     "moxie-zip": "0.0.4",
     "npm-run-all": "^4.1.5",
-    "patch-package": "^7.0.0",
+    "patch-package": "^8.0.0",
+    "postcss": "^8.4.35",
+    "postcss-preset-env": "^9.4.0",
```
Comment (TheSpyder) [inline package.json:77]: This should be a transitive dependency from stylelint? I don't think we use it directly?

Comment (ArvinJ-H) [reply package.json:77]: this was added to remove the warnings caused by `postcss-preset-env` during install

**Review by ArvinJ-H — APPROVED** [review]

File: package.json (hunk 2 above)
Comment (TheSpyder) [reply package.json]: That's not an acceptable reason. We always need to question weird stuff especially when it comes to package management. Please check this sort of thing in detail next time.

I dug in further and it's because `less-plugin-preset-env` has `postcss-preset-env` as a peer dependency. No packages specify it as a direct dependency.

File: modules/oxide-icons-default/package.json [hunk 3]
```diff
@@ -25,13 +25,6 @@
...
   "devDependencies": {
-    "gulp": "^4.0.2",
-    "gulp-clean": "^0.4.0",
-    "gulp-clean-css": "^4.3.0",
-    "gulp-connect": "^5.7.0",
-    "gulp-less": "^5.0.0",
-    "gulp-rename": "^2.0.0",
-    "gulp-sourcemaps": "^3.0.0",
-    "gulp-stylelint": "^13.0.0"
+    "grunt": "^1.6.1"
```
Comment (TheSpyder) [inline modules/oxide-icons-default/package.json:28]: TinyMCE puts all dev dependencies used by multiple projects at the monorepo root
```suggestion
```

File: modules/oxide/package.json [hunk 4]
```diff
@@ -36,6 +37,16 @@
...
   },
   "devDependencies": {
-    "gulp": "^4.0.2"
+    "connect": "^3.7.0",
+    "connect-livereload": "^0.6.1",
+    "grunt": "^1.6.1",
+    "grunt-contrib-clean": "^2.0.1",
+    "grunt-contrib-cssmin": "^5.0.0",
+    "grunt-contrib-less": "^3.0.0",
+    "grunt-stylelint": "^0.20.1",
```
Comment (TheSpyder) [inline modules/oxide/package.json:46]:
```suggestion
```

If you look closely at the `Gruntfile`, these dependencies aren't respected even if they served any purpose 😉

File: modules/oxide/package.json [hunk 5]
```diff
@@ -36,6 +37,16 @@
...
+    "less": "^4.2.0",
+    "stylelint": "^16.6.1",
+    "stylelint-config-recommended-less": "^3.0.1",
+    "stylelint-order": "^6.0.4"
```
Comment (TheSpyder) [inline modules/oxide/package.json:50]:
```suggestion
```
Similar to the icons package, these don't need to be specified twice

(nor does grunt, please remove that too)

File: modules/oxide/src/less/theme/components/dialog/dialog.less [hunk 6]
```diff
@@ -507,123 +510,124 @@
...
-    transition: visibility 0s linear .3s, opacity .3s ease;
-  }
+    .tox-dialog-dock-transition {
+      transition: visibility 0s linear .3s, opacity .3s ease;
+    }
 
-  .tox-dialog-dock-transition.tox-dialog-dock-fadein {
-    transition-delay: 0s;
-  }
-}
```
Comment (TheSpyder) [inline modules/oxide/src/less/theme/components/dialog/dialog.less:562]: This nesting change is incorrect, the styles below are top-level `.tox` styles

File: modules/oxide/src/less/theme/components/dialog/dialog.less [hunk 7]
```diff
@@ -507,123 +510,124 @@
     iframe {
       border-radius: @dialog-iframe-bordered-border-radius;
     }
-  }
```
Comment (TheSpyder) [inline modules/oxide/src/less/theme/components/dialog/dialog.less:510]: This nesting change is incorrect

File: modules/oxide/src/less/theme/components/form/color.less [hunk 8]
```diff
@@ -16,7 +16,7 @@
     }
 
     span {
-      border-color: rgba(red(@color-black), green(@color-black), blue(@color-black), .2);
+      border-color: rgba(red(@color-black) green(@color-black) blue(@color-black) / 20%);
```
Comment (TheSpyder) [inline modules/oxide/src/less/theme/components/form/color.less:19]: I don't know if I'll ever get used to the new syntax. This still looks to me like it's dividing the blue by 20% (I know it's not).

File: modules/oxide/src/less/theme/components/pop/pop.less [hunk 9]
```diff
@@ -63,15 +63,15 @@
   .tox-pop::before,
   .tox-pop::after {
     border-style: solid;
-    content: '';
+    content: 0;
```
Comment (TheSpyder) [inline modules/oxide/src/less/theme/components/pop/pop.less:66]: This looks wrong?

File: modules/oxide/src/less/theme/components/toolbar-button/toolbar-button.less [hunk 10]
```diff
@@ -34,10 +35,12 @@
 
 @toolbar-button-active-background-color: @toolbar-button-enabled-background-color;
 @toolbar-button-active-border: @toolbar-button-enabled-border;
-@toolbar-button-active-box-shadow: @toolbar-button-box-shadow; // set to `0 1px 1px 0 rgba(0, 0, 0, .1) inset` for a small push in effect
+@toolbar-button-active-box-shadow: @toolbar-button-box-shadow;
+/* set to `0 1px 1px 0 rgba(0, 0, 0, .1) inset` for a small push in effect */
```
Comment (TheSpyder) [inline modules/oxide/src/less/theme/components/toolbar-button/toolbar-button.less:39]: Comment moved down instead of up
```suggestion
/* set to `0 1px 1px 0 rgba(0, 0, 0, .1) inset` for a small push in effect */
@toolbar-button-active-box-shadow: @toolbar-button-box-shadow;
```

File: modules/oxide/src/less/theme/components/toolbar-button/toolbar-button.less [hunk 11]
```diff
@@ -34,10 +35,12 @@
...
 @toolbar-button-active-icon-color: @toolbar-button-enabled-icon-color;
 @toolbar-button-active-text-color: @toolbar-button-enabled-text-color;
-@toolbar-button-active-transform: none; // Set to `translateY(1px)` for a small push down effect
+@toolbar-button-active-transform: none;
+/* Set to `translateY(1px)` for a small push down effect */
```
Comment (TheSpyder) [inline modules/oxide/src/less/theme/components/toolbar-button/toolbar-button.less:43]: Comment moved down instead of up
```suggestion
/* Set to `translateY(1px)` for a small push down effect */
@toolbar-button-active-transform: none;
```

File: modules/oxide/src/less/theme/components/toolbar-button/toolbar-split-button.less [hunk 12]
```diff
@@ -26,7 +26,6 @@
     z-index: 1; // Ensure focus outline is on top of other buttons
 
     &::after {
-      pointer-events: none; // Prevent the focus outline from opening the menu
```
Comment (TheSpyder) [inline modules/oxide/src/less/theme/components/toolbar-button/toolbar-split-button.less:29]: Why was this removed?

File: modules/oxide/src/less/theme/content/resize-handles/resize-handle.less [hunk 13]
```diff
@@ -61,7 +61,7 @@
 
   .mce-resize-helper {
     background: #555;
-    background: rgba(0, 0, 0, .75);
+    background: rgba(0 0 0 75%);
```
Comment (TheSpyder) [inline modules/oxide/src/less/theme/content/resize-handles/resize-handle.less:64]: Doesn't this need a `/`?
```suggestion
    background: rgba(0 0 0 / 75%);
```

File: package.json [hunk 14]
```diff
@@ -38,53 +38,60 @@
...
+    "stylelint-less": "^3.0.1",
+    "stylelint-order": "^6.0.4",
+    "terser": "^5.28.1",
     "ts-loader": "^9.2.6",
     "tsconfig-paths-webpack-plugin": "^4.0.1",
     "tslib": "^2.0.0",
     "twemoji": "^14.0.2",
-    "typescript": "^4.8.3",
-    "webpack": "^5.74.0",
+    "typescript": "^5.3.3",
```
Comment (TheSpyder) [inline package.json:92]: um, woah woah woah. 

Aside from the fact that you shouldn't go around upgrading every dependency in a random PR, TS5 is a big change. It's good to know that the editor still compiles, I'm sure someone said we had errors after upgrading. Maybe Microsoft fixed them since 5 launched.

- [ ] Revert any change not directly related to stylelint or grunt

File: modules/oxide/src/less/skins/ui/tinymce-5/skin.less [hunk 15]
```diff
@@ -178,7 +178,7 @@
   .tox-statusbar__wordcount {
     &:hover,
     &:focus {
-      @{statusbar-not-disabled-selector} {
+      &:not(:disabled):not([aria-disabled=true]) {
```
Comment (TheSpyder) [inline modules/oxide/src/less/skins/ui/tinymce-5/skin.less:181]: You moved the definition of this variable into this file, but aren't using it? What's the story here? Same in `statusbar.less`

**Review by TheSpyder — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (ArvinJ-H) [reply package.json]: yes and `postcss-preset-env` requires `postcss` to function, and it doesn't seem to pull in as a peer dep automatically.

File: modules/oxide/src/less/theme/components/pop/pop.less (hunk 9 above)
Comment (ArvinJ-H) [reply modules/oxide/src/less/theme/components/pop/pop.less:66]: reverted, issue caused by stylelint config previously

File: modules/oxide/src/less/theme/components/toolbar-button/toolbar-split-button.less (hunk 12 above)
Comment (ArvinJ-H) [reply modules/oxide/src/less/theme/components/toolbar-button/toolbar-split-button.less:29]: sorry, deleted by mistake.

File: modules/oxide/README.md [hunk 16]
```diff
@@ -7,4 +7,5 @@ Visit the [TinyMCE 7 documentation](https://www.tiny.cloud/docs/tinymce/7/creati
 The build process uses [Node](http://nodejs.org/) and [Gulp](http://gulpjs.com/). Make sure you have both installed before you continue.
 
 1. Install dependencies with `yarn`.
-2. To view the skins, run `npm start` in this repository and see the oxide-test-data repository for more information.
+2. Build the skins with `yarn build`.
+3. To view the skins, run `yarn start` in this repository and go to `localhost:4000`.
```
Comment (TheSpyder) [inline modules/oxide/README.md:11]:
```suggestion
3. To view the skins, run `yarn start` in this repository and go to `http://localhost:4000`.
```

**Review by TheSpyder — APPROVED** [review]
much nicer 👍

File: modules/oxide/.stylelintrc [hunk 17]
```diff
@@ -1,12 +1,30 @@
...
+    "no-descending-specificity": null,
+    "no-invalid-double-slash-comments": null,
+    "number-max-precision": null,
+    "selector-attribute-quotes": null,
+    "selector-not-notation": null,
+    "shorthand-property-no-redundant-values": null,
+    "value-keyword-case": null,
+    "value-no-vendor-prefix": null,
+    "comment-empty-line-before": null,
+    "comment-no-empty": null
```
[the remark this answers is not in the document]
Comment (ArvinJ-H) [reply modules/oxide/.stylelintrc:28]: they don't seems to be needed anymore, maybe overridden by the new `stylelint-config-recommended-less`. Should we leave it for potential error down the road?

[the remark this answers is not in the document]
Comment (tiny-ben-tran) [reply modules/oxide/.stylelintrc:28]: IMO less code is the best. We can worry about them later if needed

**Review by MitchC1999 — COMMENTED** [review]
@ArvinJ-H Excellent work, looking very clean now. Thanks for sorting this out

**Review by shanmen-tiny — APPROVED** [review]
Do we still need the `oxide-icons-default/gulpfile.js`?

**Comment by TheSpyder** [comment]
We do not

- [ ] delete any remaining references to gulp including `gulpfile.js` files

File: .changes/unreleased/tinymce-TINY-10116-2024-07-22.yaml [hunk 18]
```diff
@@ -0,0 +1,6 @@
+project: tinymce
+kind: Changed
+body: Replaced Gulp with Grunt in Oxide for package building processes.
```
Comment (MitchC1999) [inline .changes/unreleased/tinymce-TINY-10116-2024-07-22.yaml:3]:
```suggestion
body: Removed Gulp dependency entirely and replaced with Grunt in all modules, finally including Oxide and oxide-icons-default.
```

File: modules/oxide/package.json [hunk 19]
```diff
@@ -36,6 +37,8 @@
     "prismjs": "^1.27.0"
   },
   "devDependencies": {
-    "gulp": "^4.0.2"
+    "connect": "^3.7.0",
+    "connect-livereload": "^0.6.1",
+    "stylelint-config-recommended-less": "^3.0.1"
```
Comment (MitchC1999) [inline modules/oxide/package.json:42]: I think these dependencies should be moved to the workspace package.json

File: .changes/unreleased/tinymce-TINY-10636-2024-07-22.yaml [hunk 20]
```diff
@@ -0,0 +1,6 @@
+project: tinymce
+kind: Improved
+body: Updated Stylelint to the latest version
```
Comment (MitchC1999) [inline .changes/unreleased/tinymce-TINY-10636-2024-07-22.yaml:3]:
```suggestion
body: Updated Stylelint to version 16.6.
```
The latest version is 16.7 as of last week.

File: .changes/unreleased/tinymce-TINY-10116-2024-07-22.yaml (hunk 18 above)
Comment (ArvinJ-H) [reply .changes/unreleased/tinymce-TINY-10116-2024-07-22.yaml:3]:
```suggestion
body: Removed Gulp and replaced it with Grunt in `Oxide` and `oxide-icons-default` modules.
```

File: modules/oxide/package.json (hunk 19 above)
Comment (ArvinJ-H) [reply modules/oxide/package.json:42]: these are module-specific dependencies, shouldn't they stay in module `package.json`

Comment (ltrouton) [reply modules/oxide/package.json:42]: I believe it is fine to leave this as is as like Arvin said, these dev dependencies are specific to the module

Comment (TheSpyder) [reply modules/oxide/package.json:42]: The open source project puts all dev dependencies at the top level.

However I stopped doing that for the premium monorepo; initially because yarn 4 required it but I think in general it makes sense to only put shared dev dependencies at the top level (because those are the ones we _really_ want to keep the versions matching across the codebase).

**Comment by TheSpyder** [comment]
Please delete both changelog entries in a new PR. Those were added after I approved.

The changes made here are our own development tooling, totally irrelevant to integrators. They don't belong in our public editor changelog.

If we had internal changelogs for `oxide` or the icons, that might be fine, but I don't think we do.
